### ISSUE TITLE
[Unified map] wrong navigation line between 1st and 2nd WP of route

### ISSUE THREAD

**Issue by ztNFny** [cgeo/cgeo#14882]
The line between the first and second WP of a route starts from width 0px and only slowly goes to its full width, resulting in a broken look.

![image](https://github.com/cgeo/cgeo/assets/1258173/ea7e68a5-8065-483c-839f-a8230e20c8ab)

**Comment by moving-bits** [comment]
this is something I have observed for VTM from time to time in different places as well, but not clearly reproducible for me

**Comment by ztNFny** [comment]
It's 100% reproducable for me using offline or online OSM. Only OK with google

**Comment by moving-bits** [comment]
> Only OK with google

Yes, only had this with VTM, not with Google Maps.

**Comment by devemux86** [comment]
@moving-bits I have not seen this before.

Maybe an OpenGL thing or something in line drawables?
It would be interesting if there a small code sample to reproduce it.

Why not use vtm-jts/PathLayer (see [PathOverlayActivity](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/PathOverlayActivity.java)) for drawing line overlays?

**Comment by moving-bits** [comment]
For implementation details and design decisions of our `GeoItemLayer` we would have to ask @eddiemuc, he implemented it as an abstraction layer for all our map implementations.

Regarding the "small code sample to reproduce it": Sadly it's not really reproducible, it just sometimes happens this way, but only infrequently. I remember having had this quite some time ago when I started with implementation of `UnifiedMap`, but even then it happened from time to time only. No clear pattern yet.

**Comment by fm-sys** [comment]
> It would be interesting if there a small code sample to reproduce it.

What we do is basically: 

```
final Style style = Style.builder()
                .strokeWidth(...)
                .strokeColor(...)
                .fillAlpha(...)
                .fillColor(...)
                .build();

drawable = new LineDrawable(/*list of 3 GeoPoints*/, style);

final VectorLayer vectorLayer = new VectorLayer(map);
vectorLayer.add(drawable);
vectorLayer.update();

```

**Comment by eddiemuc** [comment]
> It's just that `PathLayer` was implemented specifically optimized for line overlays. So most usually use and _test_ this method all this time.

I took a look, and it turns out that `PathLayer` is simply an extension of `VectorLayer`. Inside it also creates a `LineDrawable` and add it to the `VectorLayer` to draw the line. So IMHO using `PathLayer` should not really make any difference.

However, one thing I couldn't really figure out from code is the last line `mWorker.submit(0)`, where mWorker is initialized to `mWorker = new Worker(mMap);` in the constructor. This I don't call, could this be the source of the problem? What does the Worker do?

From PathLayer class:

```
public class PathLayer extends VectorLayer {

...

    public void addPoints(Collection<? extends GeoPoint> pts) {
        mPoints.addAll(pts);
        updatePoints();
    }

    private void updatePoints() {
        synchronized (this) {

            if (mDrawable != null) {
                remove(mDrawable);
                mDrawable = null;
            }

            if (!mPoints.isEmpty()) {
                mDrawable = new LineDrawable(mPoints, mStyle);
                if (mDrawable.getGeometry() == null)
                    mDrawable = null;
                else
                    add(mDrawable);
            }
        }
        mWorker.submit(0);
    }

...

}
```

From AbstractVectorLayer (which is extended by VectorLayer, which is extended by PathLayer):

```

    public AbstractVectorLayer(Map map) {
        super(map);
        mWorker = new Worker(mMap);
        mRenderer = new Renderer();
    }

```

**Comment by eddiemuc** [comment]
> See the [VectorLayerActivity](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/VectorLayerActivity.java) example.
> When making changes must call `vectorLayer.update()` which calls the worker.

Ah ok, thanks for the info. We do call `vectorLayer.update()`, so that should be fine

**Comment by eddiemuc** [comment]
Fixed with PR #15007

Root cause is that route passes polylines to VTM which contains the same Geopoint multiple times in conjunction. Apparently when VTM gets such a polyline with the same gepoint at the start multiple times, this results in a drawing error as observed in this issue.

PR #15007 fixes this by removing duplicated consecutive geopoints from all polyline objects in GeoPrimitive class.
